csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -182,11 +182,251 @@ bool InnerOuterPersistentKernelScheduler::canScheduleCompileTime(
+    auto input_tv = projectable_buffer_inputs[idx];
+    n_proj_buffer_uses[idx] = (int)input_tv->uses().size();
+    for (auto buffer : persistent_buffers_original) {
+      if (DependencyCheck::isDependencyOf(input_tv, buffer)) {
+        n_proj_buffer_uses[idx] += (int)buffer->uses().size() - 1;
+      }
+    }
+  }
+
+  // assume the fusion is scheduled with a two-step approach, the first step
+  // finishes the inner reduction and a partial outer reduction. The second
+  // step finishes the outer reduction. In the first step, the outer dimension
+  // is parallelized by BIDy, each thread loops over the outer dimension
+  // for n_outer_loop = ceilDiv(outer_dim_size, gdimy) times. gdimy is a
+  // heuristic para, here we assume gdimy = sm_count. For outer broadcast tv,
+  // e.g. weight tensor, it is reused in every outer loop, when calculate the
+  // latency of global memory load, 1 is used instead of n_outer_loop. For other
+  // tensors, each loop is loading a different row, so n_outer_loop is used.
+  std::vector<int> buffer_latency(n_buffer_inputs, 0);
+  int sm_count =
+      (int)at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
+  int n_outer_loop = [&sm_count, &reduction_tvs, &runtime_info]() -> int {
+    for (auto buffer : reduction_tvs) {
+      if (scheduler_utils::isFastestDimReduction(buffer)) {
+        continue;
+      }
+      int64_t outer_dim_size = -1;
+      for (auto id : buffer->getMaybeRFactorDomain()) {
+        if (id->isReduction()) {
+          auto id_size =
+              runtime_info.expressionEvaluator().evaluate(id->extent());
+          NVF_ERROR(id_size.hasValue(), "Could not infer outer dim size.");
+          if (outer_dim_size == -1) {
+            outer_dim_size = id_size.as<int64_t>();
+          } else {
+            outer_dim_size *= id_size.as<int64_t>();
+          }
+        }
+      }
+      int gdimy = sm_count;
+      int n_loops = (int)ceilDiv(outer_dim_size, gdimy);
+      return n_loops;
+    }
+    NVF_ERROR(false, "Could not infer outer loop count.");
+    return 1;
+  }();
+
+  // set latency for each buffer.
+  // Assume load from gmem to smem has a latency of 100, load from smem to
+  // register has a latency of 10. Total latency is proportional to data type
+  // size. Reused in each iteration if the buffer is an outer broadcast tensor.
+  for (auto idx = 0; idx < n_buffer_inputs; idx++) {
+    int latency = 0;
+    auto input_buffer = projectable_buffer_inputs.at(idx);
+    bool is_reused = std::any_of(
+        outer_broadcast_tvs.begin(),
+        outer_broadcast_tvs.end(),
+        [&input_buffer](TensorView* tv) {
+          return DependencyCheck::isDependencyOf(input_buffer, tv);
+        });
+    // for each load from gmem to smem, latency += 100
+    latency += 100 * (is_reused ? 1 : n_outer_loop);
+    // for each load from smem to regs, latency += 10
+    latency += 10 * n_outer_loop * n_proj_buffer_uses.at(idx);
```

Comment:
This look a bit different from what we have in the doc.

here it is really
((is_outer_broadcast_tvs? 1 : S) * 100 + S * N * 10) * dtypeSize

Should the n_outer_loop in line 373 be replaced?

Reply:
The code is correct, I will revise the doc.
The code structure is:
```
Tb in smem = load from gmem
for i = 0 , n_outer_loop:
    load Tb to register
```
The outer broadcast tv only needs to do one load from gmem to smem, but it still needs to be loaded from gmem to register when it is being used in the loop.